pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/generator/StageDirective.java

```diff
@@ -0,0 +1,152 @@
+    private List<AbstractDirective> directives = new ArrayList<>();
+    private String name;
+    private int contentType;
+
+    @DataBoundConstructor
+    public StageDirective(List<AbstractDirective> directives, @Nonnull String name, int contentType) {
+        if (directives != null) {
+            this.directives.addAll(directives);
+        }
+        this.name = name;
+        this.contentType = contentType;
+    }
+
+    @Nonnull
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * What the content of this stage is - currently steps or parallel
+     * @return 0 for steps, 1 for parallel.
+     */
+    public int getContentType() {
+        return contentType;
+    }
+
+    @Nonnull
+    public List<AbstractDirective> getDirectives() {
+        return directives;
+    }
+
+    @Extension
+    public static class DescriptorImpl extends DirectiveDescriptor<StageDirective> {
+        @Override
+        @Nonnull
+        public String getName() {
+            return "stage";
+        }
+
+        @Override
+        @Nonnull
+        public String getDisplayName() {
+            return "Stage";
+        }
+
+        @Override
+        @Nonnull
+        public List<Descriptor> getDescriptors() {
+            List<Descriptor> descriptors = new ArrayList<>();
+            descriptors.add(Jenkins.getActiveInstance().getDescriptorByType(AgentDirective.DescriptorImpl.class));
+            descriptors.add(Jenkins.getActiveInstance().getDescriptorByType(InputDirective.DescriptorImpl.class));
+            descriptors.add(Jenkins.getActiveInstance().getDescriptorByType(EnvironmentDirective.DescriptorImpl.class));
+            descriptors.add(Jenkins.getActiveInstance().getDescriptorByType(OptionsDirective.DescriptorImpl.class));
+            descriptors.add(Jenkins.getActiveInstance().getDescriptorByType(WhenDirective.DescriptorImpl.class));
+            descriptors.add(Jenkins.getActiveInstance().getDescriptorByType(ToolsDirective.DescriptorImpl.class));
+            descriptors.add(Jenkins.getActiveInstance().getDescriptorByType(PostDirective.DescriptorImpl.class));
+
+            return descriptors;
+        }
+
+        public FormValidation doCheckName(@QueryParameter String value) {
+            if (StringUtils.isEmpty(value)) {
+                return FormValidation.error("Stage name must be provided.");
+            } else {
+                return FormValidation.ok();
+            }
+        }
+
+        @Override
+        @Nonnull
+        public String toGroovy(@Nonnull StageDirective directive) {
+            StringBuilder result = new StringBuilder("stage(");
+            result.append("'").append(directive.name).append("') {\n");
+            switch (directive.contentType) {
+                case 0:
+                    result.append("steps {\n");
+                    result.append("// One or more steps need to be included within the steps block.\n");
```

Comment:
Everything gets auto-indented magically via the same method we use to indent the Jenkinsfiles translated from the editor - which wants everything to be left-aligned when passed to it so it can figure out the right indent properly.